enclave/core/sgx/calls.c

```diff
@@ -236,126 +237,222 @@ oe_result_t oe_register_ecall_function_table(
     // whose first field is an oe_result_t. The function is expected
     // to fill this field with the status of the ecall.
-    result = *(oe_result_t*)output_buffer;
+    result = *(oe_result_t*)enc_output_buffer;
 
-    if (result == OE_OK)
+    if (OE_OK == result)
     {
         // Copy outputs to host memory.
-        memcpy(args.output_buffer, output_buffer, output_bytes_written);
+        memcpy(host_output_buffer, enc_output_buffer, enc_output_bytes_written);
 
         // The ecall succeeded.
-        args_ptr->output_bytes_written = output_bytes_written;
-        args_ptr->result = OE_OK;
+        *host_output_bytes_written = enc_output_bytes_written;
+        *host_result = OE_OK;
     }
 
 done:
     if (buffer)
+    {
         oe_free(buffer);
+    }
 
+    return result;
+} /* _handle_ecall_function */
+
+static oe_result_t _handle_call_enclave_function(uint64_t arg_in)
+{
+    oe_call_enclave_function_args_t args, *args_ptr;
+    oe_result_t result = OE_OK;
+
+    // Ensure that args lies outside the enclave.
+    if (!oe_is_outside_enclave(
+            (void*)arg_in, sizeof(oe_call_enclave_function_args_t)))
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    // Copy args to enclave memory to avoid TOCTOU issues.
+    args_ptr = (oe_call_enclave_function_args_t*)arg_in;
+    args = *args_ptr;
+
+    result = _handle_ecall_function(
+        args.table_id,
+        args.function_id,
+        (const uint8_t*)args.input_buffer,
+        args.input_buffer_size,
+        args.output_buffer,
+        args.output_buffer_size,
+        &(args_ptr->output_bytes_written),
+        &(args_ptr->result));
+
+done:
     return result;
 }
 
+static oe_result_t _handle_synchronous_switchless_ecall(
+    oe_switchless_synchronous_ecall_t* ecall)
+{
+    oe_result_t result = OE_UNEXPECTED;
+
+    // Ensure that args lies outside the enclave.
+    if (!oe_is_outside_enclave(
+            (void*)ecall, sizeof(oe_switchless_synchronous_ecall_t)))
+    {
+        OE_RAISE(OE_INVALID_PARAMETER);
+    }
+
+    result = _handle_ecall_function(
+        ecall->table_id,
+        ecall->function_id,
+        ecall->input_buffer,
+        ecall->input_buffer_size,
+        ecall->output_buffer,
+        ecall->output_buffer_size,
+        &(ecall->output_bytes_written),
+        &(ecall->result));
+
+done:
+    __atomic_clear(&(ecall->lock), __ATOMIC_RELEASE);
+
+    return result;
+} /* _handle_synchronous_switchless_ecall */
+
+static oe_switchless_state_t _get_switchless_state(oe_switchless_t* switchless)
+{
+    oe_switchless_state_t state;
+    state = __atomic_load_n(&(switchless->state), __ATOMIC_ACQUIRE);
+    return state;
+} /* _get_switchless_state */
+
+static void _handle_launch_enclave_worker(uint64_t arg_in)
+{
+    oe_switchless_t* switchless = (oe_switchless_t*)arg_in;
+
+    /* loop until the state changes to stopping */
+    while (OE_SWITCHLESS_STATE_STOPPING != _get_switchless_state(switchless))
```

Comment:
Without pointer validation this loop condition can be used to leak information.

Reply:
Validation added 659cbd3